File: talleres/taller-1/code/experiment.py
import time


def is_sorted(seq):
    """Checks if the sequence is sorted.

    Args:
        seq: a sequence of elements

    Returns:
        True if the sequence is sorted, False otherwise
    """
    f = True
    for i in range(len(seq) - 1):
        f = f and not (seq[i + 1] < seq[i])
    return f


def do_experiment(seq, sorting_algorithm):
    """Sorts the sequence using the given sorting algorithm. It runs the
    experiment for 10 times and returns the average time.

    Args:
        seq: a sequence of elements
        sorting_algorithm: a sorting algorithm that takes a sequence as input
        and returns a sorted sequence

    Returns:
        [is_sorted, avg_time]
    """
    run_times = 10
    time_total = 0
    sorted = True
    for i in range(run_times):
        copy_seq = seq.copy()
        start = time.time()
        sorting_algorithm(copy_seq)
        end = time.time()
        sorted = sorted and is_sorted(copy_seq)
        time_total += float(end - start)
    return [sorted, time_total / float(run_times)]

File: talleres/taller-2/code/experiment.cpp
#include "experiment.h"
#include <chrono>


/**
 * @brief The function runs an experiment that measures the time it takes to
 * compute the reversed binary representation of a number through the passed
 * algorithm. The function runs the experiment multiple times and returns the
 * average time.
 * 
 * @param x the number to find its reversed binary representation
 * @param function the algorithm to use
 * @param run_times the number of times to run the experiment. Default is 10.
 * @return (double) the average time it takes in nanoseconds (10^-9 s)
 */
double doExperiment(
    unsigned int x, unsigned int (*function)(unsigned int), unsigned int run_times
) {
    if (run_times < 0) {
        return -1;
    }
    double time_total = 0;
    for (unsigned int i = 0; i < run_times; i++) {
        auto start = std::chrono::steady_clock::now();
        int y = function(x);
        auto end = std::chrono::steady_clock::now();
        time_total += std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count();
    }
    return time_total / run_times;
}
